python/lsst/drp/tasks/assemble_coadd.py

```diff
@@ -441,19 +454,16 @@ def runQuantum(self, butlerQC, inputRefs, outputRefs):
             log.warning("doMaskBrightObjects is set to True, but brightObjectMask not loaded")
         self.processResults(retStruct.coaddExposure, inputData["brightObjectMask"], outputDataId)
 
-        if self.config.doWriteArtifactMasks and supplementaryData:  # branch to see if running CompareWarp.
-            dataIds = [ref.dataId for ref in inputs.warpRefList]
-            psfMatchedDataIds = [ref.dataId for ref in supplementaryData.warpRefList]
-
-            if dataIds != psfMatchedDataIds:
-                supplementaryData.warpRefList = reorderAndPadList(
-                    supplementaryData.warpRefList, psfMatchedDataIds, dataIds
-                )
+        if self.config.doWriteArtifactMasks:
+            artifactMasksRefList = reorderAndPadList(
+                outputRefs.artifactMasks,
+                [ref.dataId for ref in outputRefs.artifactMasks],
+                [ref.dataId for ref in inputs.warpRefList],
+            )
             for warpRef, altMask, outputRef in zip(
-                inputs.warpRefList, retStruct.altMaskList, outputRefs.artifactMasks
+                inputs.warpRefList, retStruct.altMaskList, artifactMasksRefList, strict=True
             ):
-                if warpRef is None:
-                    continue
+                assert warpRef.dataId == outputRef.dataId, "DataId mismatch between warp and artifactMask"
```

Comment:
You're absolutely right. I should condition on `outputRef`.

Reply:
I'm back to saying that checking for `None` and `continue`ing is unnecessary.

First, these data handles are from the time QG was generated and they know nothing about which quanta actually succeeded. This is particularly true with `outputRef`, which would only get reordered and never padded. Second, the `altMaskList` may be padded, and yet a check is not warranted. If the padded value (it's a `dict` with specific schema of empty list) is passed, it simply produces a trivial mask with all 0s (these are just the additional mask that needs to be applied). It'd be cleaner to persist it than skip it, since the storage cost to save a mask is low and similar masks would be persisted if there were no artifacts whatsoever. If we attempt to not persist this for padded quanta, then we would have to special case in `assemble_cell_coadd` and it would not be worth the trouble IMO.

P.S.: I was concerned why the previously failing quanta were passing with this branch setup even before I added this guard, and I have an explanation now.

Reply:
Ah I see your point.  All output refs will be defined, so there won't be padding there.  That makes sense.

Reply:
For the future us, if the corresponding PSF-matched warp is missing for a warp, then the `artifact_mask` is set to `NO_DATA` everywhere, which will automatically avoid a warp without its PSF-match counterpart from entering into a coadd.